src/timers.c

```diff
@@ -266,32 +275,39 @@ void handle_alarm(void) {
    */
 
   /* It's possible that alarms are blocked when this function is
-   * called, if so, increment alarm_pending and exit swiftly.
+   * called.  If so, increment alarms_pending and exit swiftly.
    */
-  while (nalarms) {
-    nalarms = 0;
 
-    if (!alarms_blocked) {
-      int alarm_elapsed, new_timeout;
-      time_t now;
+  if (alarms_blocked == TRUE) {
+    alarms_pending++;
+    return;
```

Comment:
If alarms are blocked, we shouldn't enter the `while` loop at all.  Previously, we'd enter the `while` loop, set `nalarms = 0`, and _then_ check whether alarms were blocked.  If they were, then perhaps the `SIGALRM` signal which triggered this function call might not be properly handled, and in fact may have been "forgotten" due to the `nalarms = 0` setting.  Perhaps that might be the cause of some of the symptoms/behaviors reported in this issue?